File: Data Structures & Algorithms/longest-palindromic-substring/submission-1.java
class Solution {
    private String expand(String s, int start, int end){
        while(start > 0 && end < s.length() - 1){
            if(s.charAt(start - 1) == s.charAt(end + 1)){
                start--;
                end++;
            }
            else break;
        }
        return s.substring(start, end + 1);
    }

    public String longestPalindrome(String s) {
        if(s.length() == 1){
            return s;
        }
        
        String res = "";
        for(int i = 0; i < s.length() - 1; ++i){
            String first = expand(s, i, i), second = (s.charAt(i) == s.charAt(i + 1)? expand(s, i, i + 1) : "");
            if(res.length() < first.length()){
                res = first;
            }
            if(res.length() < second.length()){
                res = second;
            }
        }
        return res;
    }
}

File: Data Structures & Algorithms/longest-palindromic-substring/submission-1.py
class Solution:
    def expand_string(self, s: str, l: int, r: int) -> tuple[int, int]:
        while l >= 0 and r < len(s) and s[l] == s[r]:
            l -= 1
            r += 1
        return l+1, r

    def longestPalindrome(self, s: str) -> str:
        if len(s) <= 1:
            return s
        
        start, end = 0, 0
        for i in range(len(s)):
            l1, r1 = self.expand_string(s, i, i)
            if (r1 - l1) > (end - start):
                start, end = l1, r1
            
            l2, r2 = self.expand_string(s, i, i+1)
            if (r2 - l2) > (end - start):
                start, end = l2, r2

        return s[start:end]